samples/powerplay/src/main/cpp/PowerPlayMultiPlayer.cpp

```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+#include <android/log.h>
+#include "PowerPlayMultiPlayer.h"
+
+static const char *TAG = "PowerPlayMultiPlayer";
+
+using namespace oboe;
+using namespace parselib;
+
+PowerPlayMultiPlayer::PowerPlayMultiPlayer() = default;
+
+PowerPlayMultiPlayer::~PowerPlayMultiPlayer() = default;
+
+void PowerPlayMultiPlayer::MyPresentationCallback::onPresentationEnded(oboe::AudioStream *oboeStream) {
+    __android_log_print(ANDROID_LOG_INFO, TAG,
+                        "==== MyPresentationCallback() called with gain %f",
+                        mParent->getGain(0)
+    );
+}
+
+void PowerPlayMultiPlayer::setupAudioStream(int32_t channelCount, oboe::PerformanceMode performanceMode) {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "setupAudioStream()");
+    mChannelCount = channelCount;
+
+    openStream(performanceMode);
+}
+
+bool PowerPlayMultiPlayer::openStream(oboe::PerformanceMode performanceMode) {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "openStream()");
+
+    // Use shared_ptr to prevent use of a deleted callback.
+    mDataCallback = std::make_shared<MyDataCallback>(this);
+    mErrorCallback = std::make_shared<MyErrorCallback>(this);
+    mPresentationCallback = std::make_shared<MyPresentationCallback>(this);
+
+    // Create an audio stream
+    AudioStreamBuilder builder;
+    builder.setChannelCount(mChannelCount);
+    builder.setDataCallback(mDataCallback);
+    builder.setErrorCallback(mErrorCallback);
+    builder.setPresentationCallback(mPresentationCallback);
+    builder.setFormat(AudioFormat::Float);
+    builder.setSampleRate(48000);
+    builder.setPerformanceMode(performanceMode);
+    builder.setFramesPerDataCallback(128);
+    builder.setSharingMode(SharingMode::Exclusive);
+    builder.setSampleRateConversionQuality(SampleRateConversionQuality::Medium);
+
+    Result result = builder.openStream(mAudioStream);
+    if (result != Result::OK) {
+        __android_log_print(
+                ANDROID_LOG_ERROR,
+                TAG,
+                "openStream failed. Error: %s", convertToText(result));
+        return false;
+    }
+
+    constexpr int32_t kBufferSizeInBursts = 2; // Use 2 bursts as the buffer size (double buffer)
+    result = mAudioStream->setBufferSizeInFrames(mAudioStream->getFramesPerBurst() * kBufferSizeInBursts);
+    if (result != Result::OK) {
+        __android_log_print(
+                ANDROID_LOG_WARN,
+                TAG,
+                "setBufferSizeInFrames failed. Error: %s", convertToText(result)
+        );
+    }
+
+    mSampleRate = mAudioStream->getSampleRate();
+    return true;
+}
+
+
+void PowerPlayMultiPlayer::triggerUp(int32_t index) {
+    mAudioStream->pause();
+}
+
+void PowerPlayMultiPlayer::triggerDown(int32_t index, oboe::PerformanceMode performanceMode) {
+    if (index < mNumSampleBuffers) {
+        mSampleSources[index]->setPlayMode();
+    }
+
+    if (!mAudioStream) setupAudioStream(mChannelCount, performanceMode);
```

Comment:
Do we need all these lines? Like setup stream calls open.

Also, why startStream() and mAudioStream->requestStart()

Maybe should we something like

void PowerPlayMultiPlayer::triggerDown(int32_t index, oboe::PerformanceMode performanceMode) {
    if (index < mNumSampleBuffers) {
        mSampleSources[index]->setPlayMode();
    }

    if (!mAudioStream || (mAudioStream->getState() == StreamState::Closed)) {
        openStream(performanceMode);
    }

    auto state = mAudioStream->getState();
    if (mAudioStream && (state == StreamState::Open || state == StreamState::Paused)) {
        startStream();
    }
}